webrtc-c/canary/jobs/runner.groovy

```diff
@@ -0,0 +1,165 @@
+      'CANARY_USE_TURN': params.USE_TURN,
+      'CANARY_TRICKLE_ICE': params.TRICKLE_ICE,
+      'CANARY_LOG_GROUP_NAME': params.LOG_GROUP_NAME,
+      'CANARY_LOG_STREAM_NAME': "${params.RUNNER_LABEL}-${clientID}-${START_TIMESTAMP}",
+      'CANARY_CHANNEL_NAME': "${env.JOB_NAME}-${params.RUNNER_LABEL}",
+      'CANARY_CLIENT_ID': clientID,
+      'CANARY_IS_MASTER': isMaster,
+      'CANARY_DURATION_IN_SECONDS': params.DURATION_IN_SECONDS
+    ].collect({ k, v -> "${k}=${v}" })
+    def credentials = [
+        [
+            $class: 'AmazonWebServicesCredentialsBinding', 
+            accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+            credentialsId: 'CANARY_CREDENTIALS',
+            secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+        ]
+    ]
+    
+    RUNNING_NODES_IN_BUILDING++
+
+    // TODO: get the branch and version from orchestrator
+    if (params.FIRST_ITERATION) {
+        // TODO: uncomment this when this is stable
+        // deleteDir()
+
+        checkout([$class: 'GitSCM', branches: [[name: params.GIT_HASH ]],
+                  userRemoteConfigs: [[url: params.GIT_URL]]])
+
+        sh """
+            cd $WORKSPACE/webrtc-c/canary && 
+            mkdir -p build && 
+            cd build && 
+            cmake .. -DCMAKE_INSTALL_PREFIX=$WORKSPACE/webrtc-c/canary/build && 
+            make -j kvsWebrtcCanary"""
+    }
+
+    RUNNING_NODES_IN_BUILDING--
+    
+    waitUntil {
+        RUNNING_NODES_IN_BUILDING == 0
+    }
+
+    if (!isMaster) {
+      sleep viewerStartupDelayInSeconds
+    }
+    
+    withEnv(envs) {
+        withCredentials(credentials) {
+            try {
+                sh """
+                    cd $WORKSPACE/webrtc-c/canary/build && 
+                    ./kvsWebrtcCanary"""
+            } catch (FlowInterruptedException err) {
+                echo 'Aborted due to cancellation'
+                throw err
+            } catch (err) {
+                HAS_ERROR = true
+                // Ignore errors so that we can auto recover by retrying
+                unstable err.toString()
+            }
+        }
+    }
+}
+
+pipeline {
+    agent {
+        label params.MASTER_NODE_LABEL
+    }
+
+    parameters {
+        choice(name: 'AWS_KVS_LOG_LEVEL', choices: ["1", "2", "3", "4", "5"])
+        booleanParam(name: 'USE_TURN')
+        booleanParam(name: 'TRICKLE_ICE')
+        string(name: 'LOG_GROUP_NAME')
+        string(name: 'MASTER_NODE_LABEL')
+        string(name: 'VIEWER_NODE_LABEL')
+        string(name: 'RUNNER_LABEL')
+        string(name: 'DURATION_IN_SECONDS')
+        string(name: 'MIN_RETRY_DELAY_IN_SECONDS')
+        string(name: 'GIT_URL')
+        string(name: 'GIT_HASH')
+        booleanParam(name: 'FIRST_ITERATION', defaultValue: true)
+    }
+
+    stages {
+        stage('Preparation') {
+            steps {
+              echo params.toString()
+            }
+        }
+
+        stage('Build and Run') {
+            failFast true
+
+            parallel {
```

Comment:
Maybe I missed it in the script, but, how are we ensuring the master and viewer run on different nodes?

Reply:
It's defined using `MASTER_NODE_LABEL` and `VIEWER_NODE_LABEL`.

Viewer's agent will be set here, https://github.com/aws-samples/amazon-kinesis-video-streams-demos/pull/32/files/bc7fe786def5b88dc5592f1c4ed3f7581a7c70f0#diff-4358959bf091ec0bf5c150bca6fee305R116-R118

Master's agent will be set here, https://github.com/aws-samples/amazon-kinesis-video-streams-demos/pull/32/files/bc7fe786def5b88dc5592f1c4ed3f7581a7c70f0#diff-4358959bf091ec0bf5c150bca6fee305R77-R79